src/libostree/ostree-blob-reader-pem.c

```diff
@@ -0,0 +1,225 @@
+ */
+
+#include "config.h"
+
+#include "ostree-blob-reader-pem.h"
+#include "ostree-blob-reader-private.h"
+#include <string.h>
+
+enum
+{
+  PROP_0,
+  PROP_LABEL
+};
+
+struct _OstreeBlobReaderPem
+{
+  GDataInputStream parent_instance;
+
+  gchar *label;
+};
+
+static void ostree_blob_reader_pem_iface_init (OstreeBlobReaderInterface *iface);
+G_DEFINE_TYPE_WITH_CODE (OstreeBlobReaderPem, _ostree_blob_reader_pem, G_TYPE_DATA_INPUT_STREAM,
+                         G_IMPLEMENT_INTERFACE (OSTREE_TYPE_BLOB_READER,
+                                                ostree_blob_reader_pem_iface_init));
+
+static void
+ostree_blob_reader_pem_iface_init (OstreeBlobReaderInterface *iface)
+{
+  iface->read_blob = ostree_blob_reader_pem_read_blob;
+}
+
+static void
+ostree_blob_reader_pem_set_property (GObject *object, guint prop_id, const GValue *value,
+                                     GParamSpec *pspec)
+{
+  OstreeBlobReaderPem *self = OSTREE_BLOB_READER_PEM (object);
+
+  switch (prop_id)
+    {
+    case PROP_LABEL:
+      self->label = g_value_dup_string (value);
+      break;
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+      break;
+    }
+}
+
+static void
+ostree_blob_reader_pem_get_property (GObject *object, guint prop_id, GValue *value,
+                                     GParamSpec *pspec)
+{
+  OstreeBlobReaderPem *self = OSTREE_BLOB_READER_PEM (object);
+
+  switch (prop_id)
+    {
+    case PROP_LABEL:
+      g_value_set_string (value, self->label);
+      break;
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+      break;
+    }
+}
+
+static void
+ostree_blob_reader_pem_finalize (GObject *object)
+{
+  OstreeBlobReaderPem *self = OSTREE_BLOB_READER_PEM (object);
+
+  g_free (self->label);
+
+  G_OBJECT_CLASS (_ostree_blob_reader_pem_parent_class)->finalize (object);
+}
+
+static void
+_ostree_blob_reader_pem_class_init (OstreeBlobReaderPemClass *klass)
+{
+  GObjectClass *gobject_class = G_OBJECT_CLASS (klass);
+
+  gobject_class->set_property = ostree_blob_reader_pem_set_property;
+  gobject_class->get_property = ostree_blob_reader_pem_get_property;
+  gobject_class->finalize = ostree_blob_reader_pem_finalize;
+
+  /*
+   * OstreeBlobReaderPem:label:
+   *
+   * The label to filter the PEM blocks.
+   */
+  g_object_class_install_property (
+      gobject_class, PROP_LABEL,
+      g_param_spec_string ("label", "", "", "", G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY));
+}
+
+static void
+_ostree_blob_reader_pem_init (OstreeBlobReaderPem *self)
+{
+}
+
+OstreeBlobReaderPem *
+_ostree_blob_reader_pem_new (GInputStream *base, const gchar *label)
+{
+  g_assert (G_IS_INPUT_STREAM (base));
+
+  return g_object_new (OSTREE_TYPE_BLOB_READER_PEM, "base-stream", base, "label", label, NULL);
+}
+
+enum PemInputState
+{
+  PEM_INPUT_STATE_OUTER,
+  PEM_INPUT_STATE_INNER
+};
+
+#define PEM_SUFFIX "-----"
+#define PEM_PREFIX_BEGIN "-----BEGIN "
```

Comment:
OpenSSL really doesn't have something for this? A lot of new string parsing in C is pretty unfortunate.

Reply:
OpenSSL provides PEM_* functions indeed, though they don't fit in our use-case reading as many PEM blocks incrementally from a file (stream); `PEM_read` expects a single complete PEM block, while `PEM_read_bio` requires to implement a BIO method backed by GInputStream (that wouldn't be difficult, but cumbersome as it also requires to use correct memory allocators: OpenSSL vs. GLib).